sdk/python/kfp/local/executor_output_utils.py

```diff
@@ -0,0 +1,213 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utilities for reading and processing the ExecutorOutput message."""
+import json
+import os
+from typing import Any, Dict, List, Union
+
+from google.protobuf import json_format
+from google.protobuf import struct_pb2
+from kfp import dsl
+from kfp.compiler import pipeline_spec_builder
+from kfp.dsl import executor
+from kfp.pipeline_spec import pipeline_spec_pb2
+
+
+def load_executor_output(
+        executor_output_path: str) -> pipeline_spec_pb2.ExecutorOutput:
+    """Loads the ExecutorOutput message from a path."""
+    executor_output = pipeline_spec_pb2.ExecutorOutput()
+    with open(executor_output_path) as f:
+        json_format.Parse(f.read(), executor_output)
+    return executor_output
+
+
+def cast_ints_as_floats_to_ints(
+    outputs_dict: Dict[str, Any],
+    outputs_typed_int: List[str],
+) -> Dict[str, Any]:
+    """Casts output fields that are typed as NUMBER_INTEGER to a Python int.
+
+    This is required, since google.protobuf.struct_pb2.Value uses
+    number_value to represent both floats and ints. When converting
+    struct_pb2.Value to a dict/json, int will be upcast to float, even
+    if the component output specifies int.
+    """
+    for float_output_key in outputs_typed_int:
+        outputs_dict[float_output_key] = int(outputs_dict[float_output_key])
+    return outputs_dict
+
+
+def get_outputs_from_executor_output(
+    executor_output: pipeline_spec_pb2.ExecutorOutput,
+    executor_input: pipeline_spec_pb2.ExecutorInput,
+    component_spec: pipeline_spec_pb2.ComponentSpec,
+) -> Dict[str, Any]:
+    """Obtains a dictionary of output key to output value from several messages
+    corresponding to the executed a component/task."""
+    executor_output = add_type_to_executor_output(
+        executor_input=executor_input,
+        executor_output=executor_output,
+    )
+
+    # merge any parameter outputs written using dsl.OutputPath with the rest of ExecutorOutput
+    executor_output = merge_dsl_output_file_parameters_to_executor_output(
+        executor_input=executor_input,
+        executor_output=executor_output,
+        component_spec=component_spec,
+    )
+
+    # collect parameter outputs from executor output
+    output_parameters = {
+        param_name: pb2_value_to_python(value)
+        for param_name, value in executor_output.parameter_values.items()
+    }
+    # process the special case of protobuf ints
+    outputs_typed_int = [
+        output_param_name for output_param_name, parameter_spec in
+        component_spec.output_definitions.parameters.items()
+        if parameter_spec.parameter_type ==
+        pipeline_spec_pb2.ParameterType.ParameterTypeEnum.NUMBER_INTEGER
+    ]
+    output_parameters = cast_ints_as_floats_to_ints(
```

Comment:
nit: naming of `cast_ints_as_floats_to_ints` is a bit hard to follow. one possibly cleaner way is to combine this and the code block above to have some function like `casting_protobuf_numbers`.

Reply:
Updated in 958362c3353224d4bda0ad588ede1f8a406d60b3.

I also noticed that this casting is needed on the input side. [3f86da7](https://github.com/kubeflow/pipelines/pull/10325/commits/3f86da7b1a53e25c6b1c234a2460e8df084fef88) adds casting when going from the ExecutorInput to JSON. This is consistent with the remote BE behavior.

This is needed because of the case like:

```python
@dsl.component
def my_comp(x: int) -> int:
    return x
```

If `x` goes in as a `float` (default representation in protobuf Value), it will be returned as a `float` and throw a runtime exception, since it expects an `int` return. Also related to [previous comment](https://github.com/kubeflow/pipelines/pull/10302#discussion_r1428394031).